### PR TITLE
[Smart Contracts] Refactor the block assemblers part 3

### PR DESCRIPTION
**Pull request by fassadlr**
@Aprogiena, @dangershony @noescape00 guys, this PR can be reviewed commit by commit as quite a bit of stuff moved around. There are no logic changes in this code only restructuring. In this PR we finally have a proper base class called (BlockAssembler) and now Pow and Pos assemblers properly inherits off it.

The abstract base implementation is not yet entirely correct as the next step will be to implement an interface on BlockAssembler and then just inject PowMining and PosMinting with that. 

So this:
```
services.AddSingleton<IPowMining, PowMining>();
services.AddSingleton<PowBlockAssembler>();
services.AddSingleton<IPosMinting, PosMinting>();
services.AddSingleton<PosBlockAssembler>();
```
in the following PRs will become:
```
services.AddSingleton<IPowMining, PowMining>();
services.AddSingleton<IPosMinting, PosMinting>();
services.AddSingleton<IBlockBuilder, BlockBuilder>();
```

BlockBuilder will then internally be responsible to figure out how to construct the block, be it via mining, staking  or smart contracts.

### CODE REVIEW

**Comment by fassadlr** [comment]
All integration and unit tests pass. I'm busy syncing the nodes and will then after test that staking and mining still works. Will post back soon.

**Comment by fassadlr** [comment]
I confirm that staking is working as per usual.

File: src/Stratis.Bitcoin.Features.Miner/BlockTemplate.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+        public List<long> TxSigOpsCost;
+
+        public string CoinbaseCommitment;
+
+        public Money TotalFee;
+
+        public BlockTemplate()
+        {
+            this.Block = new Block();
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Miner/BlockTemplate.cs]: This is wrong mate (you will need to merge and use the factory to create a block)

File: src/Stratis.Bitcoin.Features.Miner/PosBlockAssembler.cs [hunk 2]
```diff
@@ -39,36 +39,35 @@ public override BlockTemplate Build(ChainedBlock chainTip, Script scriptPubKey)
...
         }
 
-        protected override void UpdateHeaders()
+        public override void OnUpdateHeaders()
         {
             this.logger.LogTrace("()");
 
-            base.UpdateHeaders();
-
-            var stake = new BlockStake(this.block);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Miner/PosBlockAssembler.cs]: Was this not used?

File: src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs [hunk 3]
```diff
@@ -234,16 +145,61 @@ private int ComputeBlockVersion(ChainedBlock prevChainedBlock, NBitcoin.Consensu
...
+        }
 
+        /// <summary>
+        /// Create coinbase transaction.
+        /// Set the coin base with zero money.
+        /// Once we have the fee we can update the amount.
+        /// </summary>
+        protected virtual void CreateCoinbase()
+        {
+            this.coinbase = new Transaction();
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs]: This is also not valid anymore

File: src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs [hunk 4]
```diff
@@ -234,16 +145,61 @@ private int ComputeBlockVersion(ChainedBlock prevChainedBlock, NBitcoin.Consensu
...
+        }
+
+        /// <summary>
+        /// Configures (resets) the builder to its default state 
+        /// before constructing a new block.
+        /// </summary>
+        protected void Configure()
+        {
+            this.BlockSize = 1000;
+            this.BlockTemplate = new BlockTemplate { Block = new Block(), VTxFees = new List<Money>() };
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs]: `new Block()` also not valid anymore

File: src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs [hunk 5]
```diff
@@ -666,5 +550,115 @@ private int UpdatePackagesForAdded(TxMempool.SetEntries alreadyAdded, Dictionary
...
+            : base(consensusLoop, dateTimeProvider, loggerFactory, mempool, mempoolLock, network)
+        {
+        }
+
+        /// <inheritdoc/>
+        public override BlockTemplate Build(ChainedBlock chainTip, Script scriptPubKey)
+        {
+            this.Logger.LogTrace("({0}:'{1}',{2}.{3}:{4})", nameof(chainTip), chainTip, nameof(scriptPubKey), nameof(scriptPubKey.Length), scriptPubKey.Length);
+
+            base.OnBuild(chainTip, scriptPubKey);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs]: Should it not be `this.OnBuild`?

File: src/Stratis.Bitcoin.Features.Miner/BlockTemplate.cs (hunk 1 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/BlockTemplate.cs]: Yes, I have now merged and fixed 👍

File: src/Stratis.Bitcoin.Features.Miner/PosBlockAssembler.cs (hunk 2 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/PosBlockAssembler.cs]: Nope! (Weird)

File: src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs (hunk 3 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs]: Yes, I have now merged and fixed 👍

File: src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs (hunk 4 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs]: Yes, I have now merged and fixed 👍

File: src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs (hunk 5 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/PowBlockAssembler.cs]: Fixed.

**Comment by dangershony** [comment]
Wait for tests and push?

**Pull request merged**